src/users/users.service.ts

```diff
@@ -0,0 +1,84 @@
+import { Injectable, HttpException, HttpStatus, Inject } from '@nestjs/common';
+import * as bcrypt from 'bcrypt';
+import { User } from './interface/users.interface';
+import { USERS_DAO, IUsersDAO } from '../DAO/usersDAO';
+
+export const USERS_SERVICE = 'USERS SERVICE';
+export interface IUsersService {
+  findOne(firstName: string): Promise<User | null>;
+  findByEmail(email: string): Promise<User | null>;
+  findById(id: number): Promise<User | null>;
+  findAll(): Promise<User[]>;
+  saveUser(user: User): Promise<User>;
+  saveUnregisteredUser(user: User): Promise<User>;
+  updateUser(user: User): Promise<User>;
+  removeUser(id: number): Promise<User>;
+}
+
+@Injectable()
+export class UsersService implements IUsersService {
+  constructor(@Inject(USERS_DAO) private readonly usersDAO: IUsersDAO) {}
+
+  async findOne(firstName: string): Promise<User | null> {
+    return await this.usersDAO.findByFirstName(firstName);
+  }
+
+  async findByEmail(email: string): Promise<User | null> {
+    return await this.usersDAO.findOne(email);
+  }
+
+  async findById(id: number): Promise<User | null> {
+    return await this.usersDAO.findById(id);
+  }
+
+  async findAll(): Promise<User[]> {
+    const allUsers: User[] = await this.usersDAO.find();
+    allUsers.forEach((user) => {
+      user.password = undefined;
+    });
+    return allUsers;
+  }
+
+  async saveUser(user: User): Promise<User> {
+    const hashedPassword: string = await bcrypt.hash(user.password, 10);
+    user.password = hashedPassword;
+    return await this.usersDAO.save(user);
+  }
+
+  async saveUnregisteredUser(user: User): Promise<User> {
+    const foundUser: User = await this.findByEmail(user.email);
+    if (foundUser) {
+      throw new HttpException('User exists!', 400);
+    }
+    try {
+      const unregisteredUser: User = await this.usersDAO.save(user);
+      return unregisteredUser;
+    } catch (e) {
+      throw new HttpException('Bad credentials!', 400);
+    }
+  }
+
+  async updateUser(user: User): Promise<User> {
+    const userFound: User = await this.findByEmail(user.email); //then check invite for expire and if exists by user.id from invites repository,
+    if (!userFound) {
+      throw new HttpException('User Not Found', HttpStatus.NOT_FOUND);
+    }
+    const hashedPassword: string = await bcrypt.hash(user.password, 10);
+    user.password = hashedPassword;
+    try {
+      await this.usersDAO.update(userFound.id, user);
+      return await this.findById(userFound.id);
+    } catch (e) {
+      throw new HttpException('Cannot update user!', 400);
```

Comment:
I've noticed the wrong error handling behavior. So, here we are going to throw an HTTP exception with a 400 status code. 400's errors are client errors. But here, potentially we could have a DB error (DB constrains error, DB fold, etc) and we need to respond with 500 errors and preferably to specify where exactly we've faced with the issue. Morefore, here we have two operations inside one try-catch block what will be painful to find where was the problem, you cannot definitely say that the update query crashed.

1. for each async operation we should have a separate try-catch block (of course, thee excuses possible);
2. we correctly can specify where exactly the error raced: in catch block, a possible solution could be to write some general message, e.g: "Failed to save user".

Reply:
done